python/ray/dag/py_obj_scanner.py

```diff
@@ -13,13 +12,23 @@
 else:
     import pickle  # noqa: F401
 
-from typing import List, Dict, Any, TypeVar, TYPE_CHECKING
-
-if TYPE_CHECKING:
-    from ray.dag.dag_node import DAGNode
+from typing import List, Dict, Any, TypeVar
+from ray.dag.base import DAGNodeBase
 
 T = TypeVar("T")
 
+# Used in deserialization hooks to reference scanner instances.
+_instances: Dict[int, "_PyObjScanner"] = {}
```

Comment:
Now we have a global variable for this. I have some concerns for a multi-threading env. It's normal that the user might have multiple thread creating the DAGs right?

Reply:
I think we have the same issue for previous implementation. Also because of GIL, multi-threaded accessing actually is not an issue here; but it could be an issue in other parts of Ray DAG. I do not think we need to take care of it at this stage.

Reply:
I'm not sure I understand this. This is a dict, right? Do you mean that dict operation inside python is atomic? For example, one instruction for dict operations?
If so, I'm ok with this.


> also in the code, the class object would not be shared by multiple threads

I'm not sure. `_instances` is per-module right? so it'll be shared across threads. If we have multiple threads creating DAG, is this still thread safe?

Reply:
yes, dict operation inside python is atomic

Reply:
Good to know! Thanks!